XenAdmin/Wizards/PatchingWizard/PatchingWizard.cs

```diff
@@ -380,16 +380,28 @@ private List<AsyncAction> GetCleanUpPoolUpdateActions()
         {
             if (PatchingWizard_UploadPage.AllIntroducedPoolUpdates != null && PatchingWizard_UploadPage.AllIntroducedPoolUpdates.Count > 0)
             {
-                return PatchingWizard_UploadPage.AllIntroducedPoolUpdates.Select(pu => GetCleanUpPoolUpdateAction(pu)).ToList();
+                return PatchingWizard_UploadPage.AllIntroducedPoolUpdates.Select(GetCleanUpPoolUpdateAction).ToList();
             }
 
             return new List<AsyncAction>();
         }
 
-        private AsyncAction GetCleanUpPoolUpdateAction(Pool_update poolUpdate)
+        private static AsyncAction GetCleanUpPoolUpdateAction(Pool_update poolUpdate)
         {
             return
-                new DelegatedAsyncAction(poolUpdate.Connection, Messages.REMOVE_PATCH, "", "", session => Pool_update.pool_clean(session, poolUpdate.opaque_ref));
+                new DelegatedAsyncAction(poolUpdate.Connection, Messages.REMOVE_PATCH, "", "", session =>
+                {
+                    try
+                    {
+                        Pool_update.pool_clean(session, poolUpdate.opaque_ref);
+                        if(!poolUpdate.AppliedOnHosts.Any())
+                            Pool_update.async_destroy(session, poolUpdate.opaque_ref);

```

Comment:
If we use async api calls, then we need to call PollToCompletion() (as well as assigning RelatedTask) so that the action does not get marked as completed before the api task is actually finished.